kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,782 @@
+local NAME_ERROR_CODE             = 3 -- response code 3 as "Name Error" or "NXDOMAIN"
+-- client specific error
+local CACHE_ONLY_ERROR_CODE       = 100
+local CACHE_ONLY_ERROR_MESSAGE    = "cache only lookup failed"
+local CACHE_ONLY_ANSWERS = tablex.readonly({ errcode = CACHE_ONLY_ERROR_CODE, errstr = CACHE_ONLY_ERROR_MESSAGE })
+local EMPTY_RECORD_ERROR_CODE     = 101
+local EMPTY_RECORD_ERROR_MESSAGE  = "empty record received"
+
+
+-- APIs
+
+local _M = {}
+local mt = { __index = _M }
+
+-- copy TYPE_*
+for k,v in pairs(resolver) do
+  if type(k) == "string" and k:sub(1,5) == "TYPE_" then
+    _M[k] = v
+  end
+end
+_M.TYPE_LAST = TYPE_LAST
+
+
+local TRIES_MT = { __tostring = cjson.encode, }
+
+
+local function stats_init(stats, name)
+  if not stats[name] then
+    stats[name] = {}
+  end
+end
+
+
+local function stats_count(stats, name, key)
+  stats[name][key] = (stats[name][key] or 0) + 1
+end
+
+
+local function stats_set(stats, name, key, value)
+  stats[name][key] = value
+end
+
+
+-- lookup or set TYPE_LAST (the DNS record type from the last successful query)
+local function insert_last_type(cache, name, qtype)
+  local key = "last:" .. name
+  if TYPE_TO_NAME[qtype] and cache:get(key) ~= qtype then
+    cache:set(key, { ttl = 0 }, qtype)
+  end
+end
+
+
+local function get_last_type(cache, name)
+  return cache:get("last:" .. name)
+end
+
+
+-- insert hosts into cache
+local function init_hosts(cache, path, preferred_ip_type)
+  local hosts = parse_hosts(path)
+
+  local function insert_answer(name, qtype, address)
+    if not address then
+      return
+    end
+
+    -- insert via the `:get` callback to prevent inter-process communication
+    cache:get(name .. ":" .. qtype, nil, function()
+      return {
+        ttl = LONG_LASTING_TTL,
+        expire = now() + LONG_LASTING_TTL,
+        {
+          name = name,
+          type = qtype,
+          address = address,
+          class = 1,
+          ttl = LONG_LASTING_TTL,
+        },
+      }, nil, LONG_LASTING_TTL
+    end)
+  end
+
+  for name, address in pairs(hosts) do
+    name = string_lower(name)
+
+    if address.ipv4 then
+      insert_answer(name, TYPE_A, address.ipv4)
+      insert_last_type(cache, name, TYPE_A)
+    end
+
+    if address.ipv6 then
+      insert_answer(name, TYPE_AAAA, address.ipv6)
+      if not address.ipv4 or preferred_ip_type == TYPE_AAAA then
+        insert_last_type(cache, name, TYPE_AAAA)
+      end
+    end
+  end
+
+  return hosts
+end
+
+
+-- distinguish the worker_events sources registered by different new() instances
+local ipc_counter = 0
+
+function _M.new(opts)
+  opts = opts or {}
+
+  -- parse resolv.conf
+  local resolv, err = utils.parse_resolv_conf(opts.resolv_conf, opts.enable_ipv6)
+  if not resolv then
+    log(WARN, "Invalid resolv.conf: ", err)
+    resolv = { options = {} }
+  end
+
+  -- init the resolver options for lua-resty-dns
+  local nameservers = (opts.nameservers and #opts.nameservers > 0) and
```

Comment:
Could we use `table.isempty` to replace `#` operator？

Reply:
this is init code only, so no performance pressure here, using the `#` operator is the common Lua idiom. So imho this is fine, no changes needed.

Reply:
I think that we should always write the high quality code, the code will be a good example for community.